Use a range of port numbers instead of a single port for load balancers

selmanj opened the pull request:
GCE forwarding rules allow you to specify a range of ports to be forwarded.  Since a single port can easily be represented in a range, we should update the ManageIQ abstraction as well.

selmanj commented:
@miq-bot add_label WIP

selmanj commented:
@miq-bot add_label wip

selmanj commented:
/cc @Ladas @agrare @erjohnso 

I'm unhappy with the `_pretty` methods, but unsure where to put them yet. Going to think overnight and come back to this; otherwise this is nearly ready for review.

selmanj commented:
@miq-bot remove_label wip

selmanj commented:
@Ladas This is ready for review.

Ladas commented:
@lpichler could you review the migration? Maybe it can be written in arel, or simplified?

lpichler commented on db/migrate/20160825213523_change_load_balancer_listener_port_to_port_range.rb:
```diff
@@ -0,0 +1,55 @@
+class ChangeLoadBalancerListenerPortToPortRange < ActiveRecord::Migration[5.0]
+  def change
+    reversible do |dir|
```
@selmanj is there reason for using `reversible`? can we use methods `def up` and `def down` ? as we have in other migrations ?

selmanj replied:
I didn't realize the standard was to use def up/down.  I suppose reversible is slightly less code, but if it's confusing to read I'm happy to fix.

lpichler commented on db/migrate/20160825213523_change_load_balancer_listener_port_to_port_range.rb:
```diff
@@ -0,0 +1,55 @@
+        # existing values
+        execute <<-SQL
+          ALTER TABLE load_balancer_listeners
+            ALTER COLUMN load_balancer_port SET DATA TYPE int4range
```
@Iadas @selmanj  It looks like  that there is no way for using Arel if we want to change type from int to int4range. so 👍

lpichler replied:
@selmanj 
or maybe we can use Arel, because you are changing name and type, so you can:
- add new column with range type
- convert data using arel
- remove old column 

@selmanj  what do you think about this solution ?

selmanj replied:
Going to take a deep look at arel.  I've never used this!

selmanj replied:
Can you describe some of the intent of using arel? I don't actually see many examples of how to use arel to convert data (especially with a postgres specific type constructor).  I'm not sure either that the created arel code would be easier to read than the raw SQL to be executed.

lpichler replied:
I assume that we have already created the new column  `instance_port_range` and we need to convert data from 
`instance_port` to `instance_port_range`

so it can be for up direction :

```
LoadBalancerListener.where.not(:instance_port => nil ).each do |x|
   x.update(:instance_port_range => x.instance_port..x.instance_port)
end
```
...

is it works for you ?

selmanj replied:
When you said to arel, I didn't realize you meant the SQL manipulation layer in ActiveRecord; instead I was trying to figure out how to write raw arel code.  Whoops!

What you wrote above totally works. The only difference I see is that each record is updated one at a time (i.e. multiple sql statements are run for each update). This might be slow (doesn't matter for a one-off migration), and it might have race conditions with other updates.

However given this is a change that we don't expect anyone to be using yet, I highly doubt someone will run the migration and also be manipulating loadbalancers at the same time. I'm happy to use this format above instead.

lpichler replied:
@selmanj  I am sorry for that I used word 'arel', yes you are right I meant ActiveRecord

lpichler commented:
@selmanj for data migrations it would be nice to have also specs for it.

lpichler commented:
@gtanzillo @jrafanie 
This migration is changing column name and type and with data migration.

Do we need also data migration when the origin(https://github.com/ManageIQ/manageiq/pull/10282) migration have been created before 1 month and it is just in upstream ?

selmanj commented:
@lpichler +1 on the specs, it's not clear if the migration works or not just by reading the code.  I'll see if I can add that.

jrafanie commented:
> @gtanzillo @jrafanie 
> This migration is changing column name and type and with data migration.
> 

@lpichler @selmanj  Yes, we should split the alter column (schema) from the data changes in two separate migrations.  It's really hard to write migration specs for schema + data changes.  It's much easier if the migration only does data changes.  Note, we don't test schema changes.

> Do we need also data migration when the origin(#10282) migration have been created before 1 month and it is just in upstream ?

@lpichler I don't know what you mean.  Can you elaborate?

selmanj commented:
@jrafanie Will do.

selmanj commented:
@jrafanie Just to double check, if I split this into three migrations:

1. Add new column
2. Migrate data
3. Drop old column

Then step 3 is no longer reversible, right?  In order for it to be reversible, I'd have to have the 'down' step recreate the column, and then seed the column with data from the original `port_range` field. If I do that, then step 3 actually has a data migration as well, and I'd have to test it.

jrafanie commented:
> @jrafanie Just to double check, if I split this into three migrations:
> 
> Add new column
> Migrate data
> Drop old column

Yes

> Then step 3 is no longer reversible, right? In order for it to be reversible, I'd have to have the 'down' step recreate the column, and then seed the column with data from the original port_range field. If I do that, then step 3 actually has a data migration as well, and I'd have to test it.

Well, you'd have to deal with the migrate up, (old to new column) and also in the reverse, down (new column back to old).  This should reside in the migrate data migration in step 2.

lpichler commented:
>  Do we need also data migration when the origin(#10282) migration have been created before 1 month and it is just in upstream ?

> @lpichler I don't know what you mean. Can you elaborate?

@jrafanie  We with @Ladas thought that we don't need data migration when the origin migration is relatively young and probably nobody have data in the changed column. But ok we cannot rely on it.

selmanj commented:
Updated PR, should be ready to review.

- Moved to three migrations, add new column, migrate data, remove old column
- Added spec for migration
- Used ActiveRecord code to perform migration rather than SQL

lpichler commented on db/migrate/20160830233558_migrate_ports_to_port_ranges_in_load_balancer_listener.rb:
```diff
@@ -0,0 +1,26 @@
+class MigratePortsToPortRangesInLoadBalancerListener < ActiveRecord::Migration[5.0]
+  class LoadBalancerListener < ActiveRecord::Base
```
@selmanj if there is column 'type'  in table, we also need `self.inheritance_column = :_type_disabled`
so:
```ruby
class LoadBalancerListener < ActiveRecord::Base
   self.inheritance_column = :_type_disabled
end
```

selmanj replied:
Good call, will add

lpichler commented on db/migrate/20160830233558_migrate_ports_to_port_ranges_in_load_balancer_listener.rb:
```diff
@@ -0,0 +1,26 @@
+  end
+
+  def up
+    LoadBalancerListener.where.not(:instance_port => nil).each do |x|
```
this can be wrapped with 
say_with_time("Updating LoadBalancerListener ..") do
..
end

selmanj replied:
I was wondering how to do this!  Will add

lpichler commented:
@selmanj just minor changes otherwise looks good 👍 , thanks !

Ladas commented:
@selmanj the arel comes handy, when the sql query is too complex to write it with pure active record: https://github.com/rails/arel 

It allows all kinds of magic. :-)

jrafanie commented on db/migrate/20160830233154_add_port_ranges_to_load_balancer_listener.rb:
```diff
@@ -0,0 +1,8 @@
+  def change
+    change_table :load_balancer_listeners do |t|
+      t.column :load_balancer_port_range, :int4range, :default => nil, :null => true
+      t.column :instance_port_range, :int4range, :default => nil, :null => true
```
@Fryguy @carbonin I don't think we've used int4range before, any concerns with using it?

carbonin replied:
Nope, looks like it was [introduced in 9.2](https://www.postgresql.org/docs/9.2/static/rangetypes.html) so I'm fine with it.

Fryguy replied:
@selmanj Do you *need* int4range?  I'd prefer using standard data types where we can, particularly if we don't plan on using the special operators.

selmanj replied:
Well, alternatively we could create two separate integer columns for min/max, and achieve something similar.  However if we ever want to see if two port ranges overlap, the queries get pretty nasty; using a range type in postgres means the query language itself is aware of how to perform these lookups, which not only benefits speed but also makes the actual implementation more readable.  But that's my opinion.

I don't know of any downsides (but I haven't used numranges extensively) other than the complexity in having to use Range objects in the AR model. It does tie ManageIQ closer to postgres, but I believe there is already postgres-specific stuff in there (like the jsonb datatypes).

Will accept whichever approach you prefer; what do you think?

Fryguy replied:
Yeah, I am always wary of adding *more* Postgres specific stuff, only because it makes it that much harder to move away from it if we need to.  However, I'm good with this change. :+1:

Fryguy replied:
@selmanj Minor, but I think `:default => nil, :null => true` are both default for a column, so no need to specify in the migration.

selmanj replied:
Yeah I was paranoid that without it, it wouldn't be explicit enough.  I'll remove it.

jrafanie commented on app/models/load_balancer_listener.rb:
```diff
@@ -15,4 +15,28 @@ class LoadBalancerListener < ApplicationRecord
+    else
+      # Range#last apparently ignores exclusiveness without an explicit
+      # parameter
+      "#{r.first}-#{r.last(1)[0]}"
```
I don't understand this.  What scenario gives you problems?

```
irb(main):005:0> (3000..3002).last
=> 3002

irb(main):006:0> (3000..3006).last
=> 3006
```

jrafanie replied:
???  😕 

```

irb(main):004:0> (3000..3000).last
=> 3000
irb(main):005:0> (3000..3000).first
=> 3000
```

selmanj replied:
It's when a range is created with `exclude_end` equal to true (or using the `...` operator)

`
irb(main):001:0> (3000..3000).last
=> 3000
irb(main):002:0> (3000...3001).last
=> 3001
irb(main):003:0> (3000...3001).to_a == (3000..3000).to_a
=> true
`

Unfortunately postgres seems to prefer ranges where the last point is excluded from the range.

carbonin replied:
It's not postgres. It exposes an option see: https://www.postgresql.org/docs/9.5/static/rangetypes.html

I would imagine there should be a similar option for the migration.

carbonin replied:
Also it looks like `active_record` is parsing the range exclusions correctly [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/postgresql/oid/range.rb#L20-L32)

selmanj replied:
Right, my comment is specifically to the behavior of `Range#last`. It doesn't return the last element in the set of values in the range (which is what I would expect); instead it returns the upper bound value of the range without regard to whether or not the range is exclusive or inclusive.  

In my opinion, I don't care if its range exclusive or inclusive. My comment regarding postgres preferring exclusive ranges is because even when using a constructor like `int4range(80, 80, '[]')', a SELECT on that value will return '[80, 81)'.

This is the same value and it doesn't really matter how it's stored!  I just want to make it clear that when we use `range_to_port_range_str`, it handles ranges that are both inclusive and exclusive properly.

carbonin replied:
Ah I see, it's a ruby problem.
The issue is that:
`(1...10).last == (1..10).last`
is `true`

selmanj replied:
Yeah! I wrote `last(1)` which looks confusing as all heck; I tried to write a comment above explaining why but I can update it if it's still confusing.

carbonin replied:
Ah I think using `#min`/`#max` here should fix the issue.
```ruby
irb(main):008:0> (1...10).min
=> 1
irb(main):009:0> (1...10).max
=> 9
irb(main):010:0> (1..10).min
=> 1
irb(main):011:0> (1..10).max
=> 10
```

selmanj replied:
Well, that's much clearer!  Updating code...

jrafanie commented on db/migrate/20160830233558_migrate_ports_to_port_ranges_in_load_balancer_listener.rb:
```diff
@@ -0,0 +1,26 @@
+
+  def up
+    LoadBalancerListener.where.not(:instance_port => nil).each do |x|
+      x.update(:instance_port_range => x.instance_port..x.instance_port,
```
Is this doing a range of the same value?  Such as`3000..3000`?  Why?

selmanj replied:
It is.  This is to migrate any existing values (where we only supported a single port forward).

jrafanie commented on db/migrate/20160830233558_migrate_ports_to_port_ranges_in_load_balancer_listener.rb:
```diff
@@ -0,0 +1,26 @@
+               :instance_port       => nil)
+    end
+    LoadBalancerListener.where.not(:load_balancer_port => nil).each do |x|
+      x.update(:load_balancer_port_range => x.load_balancer_port..x.load_balancer_port,
```
Same

selmanj replied:
See above

selmanj commented:
* Added `say_with_time`
* Added type qualifier to AR model

selmanj commented:
* Used `Range#min`/`Range#max` instead of more confusing `first`/`last`

chessbyte commented:
Assigned to @Fryguy because he wants to review all SQL Migrations.

Fryguy commented on app/models/load_balancer_listener.rb:
```diff
@@ -15,4 +15,26 @@ class LoadBalancerListener < ApplicationRecord
 
   virtual_total :total_vms, :vms, :uses => :vms
+
+  def load_balancer_port_range_str
```
Not sure I like the `_str` ending.  I generally don't like abbreviations, and it doesn't really provide intent, so perhaps it should be named `display_load_balancer_port_range`.  Alternately, if this is for display/UI purposes, perhaps it should live in the UI directly as a general presentation helper named `format_port_range(range)`, which could then be called in the UI like `format_port_range(load_balance_listener.load_balancer_port_range)`

selmanj replied:
Sounds good to me. That definitely seems like a better spot for the method.

Fryguy commented on app/helpers/load_balancer_helper/textual_summary.rb:
```diff
@@ -28,7 +28,8 @@ def textual_description
   def textual_listeners
     @record.load_balancer_listeners.map do |x|
-      "Load Balancer #{x.load_balancer_protocol}:#{x.load_balancer_port}, Instance #{x.instance_protocol}:#{x.instance_port}"
+      "Load Balancer #{x.load_balancer_protocol}:#{x.load_balancer_port_range_str}, Instance"\
```
Is protocol always part of the presentation string?  If so, perhaps the `load_balancer_port_range_str` should just do that (this sort of also supports my argument for a general helper method that formats port ranges in a common way).

selmanj replied:
I actually don't know; in general the standard for port ranges seems to be <START>-<END> inclusive (or with a : instead of a -) but the prepending of the protocol may or may not be familiar.  Still, it seems readable enough to me so it seems worth including it in the helper

selmanj commented:
* Moved port rendering code to helper as display code
* Add spec for display rendering

Fryguy commented on db/migrate/20160830233842_drop_ports_in_load_balancer_listener.rb:
```diff
@@ -0,0 +1,6 @@
+class DropPortsInLoadBalancerListener < ActiveRecord::Migration[5.0]
+  def change
+    remove_column :load_balancer_listeners, :load_balancer_port, :integer, :null => true, :default => nil
+    remove_column :load_balancer_listeners, :instance_port, :integer, :null => true, :default => nil
```
Same here...no need for the `:null => true, :default => nil`

selmanj replied:
Will fix

Fryguy commented:
@selmanj Had a couple more comments about the migrations, but otherwise, looks good!

selmanj commented:
* Removed verbose `:default` and `:null` options

sergio-ocon commented:
Ping @Fryguy
